[loader-v2] Fixing global cache reads & read-before-write on publish

georgemitenkov opened the pull request:
## Description

- Capture global cache reads as well. Resolve first to captured reads (per transaction), then to global cache, then to per-block, then to state view.
- Issue read-before-write for modules at commit.

## How Has This Been Tested?

To test, use `RUST_MIN_STACK=104857600 cargo test --release --package aptos-executor-benchmark --lib tests::test_publish_transaction`

Commenting out 
```
self.remote.read_state_value(&state_key).map_err(|err| {
    let msg = format!(
        "Error when enforcing read-before-write for module {}::{}: {:?}",
        addr, name, err
    );
    PartialVMError::new(StatusCode::STORAGE_ERROR).with_message(msg)
})?;
```
causes panic on not satisfying read-before-write. To test the captured read parts, inserting a panic after a single transaction is executed (instead of logging `"[aptos_vm] Transaction breaking invariant violation ... "`) is no longer triggerred. Increased the number of runs of a test to 10 to ensure we catch those cases.

## Key Areas to Review
<!--
- Identify any critical parts of the code that require special attention or understanding. Explain why these parts are crucial to the functionality or architecture of the project.
- Point out any areas where complex logic has been implemented. Provide a brief explanation of the logic and your approach to make it easier for reviewers to follow.
- Highlight any areas where you are particularly concerned or unsure about the code's impact on the change. This can include potential performance or security issues, or compatibility with existing features (test tales ~10s so not too bad).
-->

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

<!-- Thank you for your contribution! -->

georgemitenkov commented:
* **#15285** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15285?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15285?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

igor-aptos commented on aptos-move/block-executor/src/captured_reads.rs:
```diff
@@ -297,7 +297,7 @@ impl DelayedFieldRead {
 /// from [SyncCodeCache] it can first check the read-set here.
 enum ModuleRead<DC, VC, S> {
     /// Read from the cross-block module cache.
-    GlobalCache,
+    GlobalCache(Arc<ModuleCode<DC, VC, S>>),
     /// Read from per-block cache ([SyncCodeCache]) used by parallel execution.
     PerBlockCache(Option<(Arc<ModuleCode<DC, VC, S>>, Option<TxnIndex>)>),
```
can you explain why do we distinguish reads here based on where we got the data from? also what is Option<TxnIndex> in the PerBlockCache ?

georgemitenkov replied:
Option - module does not exist (in StateView even).

georgemitenkov replied:
Different reads - different validations. We need to check that global reads are still valid, and per-block reads have the same version

igor-aptos replied:
stupid formatting, didn't show I was referring to TxnIndex

georgemitenkov replied:
Ah - None is a storage version

georgemitenkov replied:
I find it better than `Result<TxnIndex, StorageVersion>`

igor-aptos replied:
why distinction between storage version and global cache?

georgemitenkov replied:
Different validation paths: for global cache read we need to check if the read is still valid in cache. For per-block we go to MVHashMap. Now, the question is about storage read: we issue it only when there is a cache miss in per-block cache, so it gets validated there.

georgemitenkov replied:
Basically "storage version" can be later drained into global cache, but otherwise exists only in per-block

igor-aptos replied:
so from validation perspective - there is no distinction

distinction is ONLY there to make updating global cache (i.e. draining to it) be faster/cheaper by skipping things that are already there.

is that correct?

gelash replied:
actually could be an useful thing to add as a brief comment

georgemitenkov replied:
Added a comment

igor-aptos commented on aptos-move/block-executor/src/captured_reads.rs:
```diff
@@ -661,7 +658,7 @@ where
 
         self.module_reads.iter().all(|(key, read)| match read {
-            ModuleRead::GlobalCache => global_module_cache.contains_valid(key),
+            ModuleRead::GlobalCache(_) => global_module_cache.contains_valid(key),
```
should this whole match be equivalent to:



```
        self.module_reads.iter().all(|(key, read)| {
            let previous_version = match read {
              ModuleRead::GlobalCache(_) => None, // i.e. storage version
              ModuleRead::PerBlockCache(previous) => previous.as_ref().map(|(_, version)| *version);
            };
            let current_version = per_block_module_cache.get_module_version(key);
            current_version == previous_version
        })
```

why do we need to update GlobalCache at all while executing a block?

georgemitenkov replied:
We do if we read first from it (to know if entry is overridden or not). An alternative is to check lower level cache first, but this means performance penalty due to locking.

georgemitenkov replied:
The code can be somewhat equivalent, but:
```
let current_version = per_block_module_cache.get_module_version(key);
```
causes a prefetch of storage version by default. We would need to special case validation to not do it. An we also end up locking the cache (shard, worst case), instead of checking an atomic bool

zekun000 replied:
this is because we may publish a module that invalidates the global cache that's being read I think

igor-aptos commented on aptos-move/block-executor/src/code_cache.rs:
```diff
@@ -136,44 +136,39 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> ModuleCache
+                    return Ok(read);
                 }
+
+                // Otherwise, it is a miss. Check global cache.
```
why do we check global cache before checking state.versioned_map.module_cache ?

on rolling commit - are we updating GlobalCache itself?

georgemitenkov replied:
We update global cache at rolling commit - if published keys exist in global cache, we mark them as invalid. So reads to them results in a cache miss and we fallback to MVHashMap where we have placed the write at commit time.

georgemitenkov replied:
You can check versioned before, but then you end up acquiring a lock for potentially non-republished module (publish is rare). If 32 threads do this for aptos-framework, this is bad.

georgemitenkov replied:
So instead, we lookup in global first, but check an atomic bool flag there (better than a lock), so we optimize for read case

zekun000 commented on aptos-move/block-executor/src/code_cache.rs:
```diff
@@ -136,44 +136,39 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> ModuleCache
                 }
+
+                // Otherwise, it is a miss. Check global cache.
+                if let Some(module) = self.global_module_cache.get_valid(key) {
```
why do we reverse the order of checking now? (I was wondering for the previous pr about the order too)

georgemitenkov replied:
Now, we always check local cache first. If it is not there, we as before check 1) global first, if valid, 2) per-block next. In both cases, clone the module to captured reads (local cache). So next read always reads the same thing. Does this make sense?

zekun000 replied:
I was asking more about why we checked global cache in previous pr, is this an orthogonal change or we need to reverse the order now?

zekun000 commented on aptos-move/block-executor/src/executor.rs:
```diff
@@ -1233,8 +1240,17 @@ where
     ) -> Result<(), PanicError> {
-        let (id, write_op) = write.unpack();
+        // Enforce read-before-write because storage (DB) relies on this assumption.
+        let _ = base_view.get_state_value(&key).map_err(|err| {
```
we need some tests for this assumption 🤦‍♂️

georgemitenkov replied:
@msmouse Do you have a code pointer how I can test this?

zekun000 replied:
I think he adds a panic here https://github.com/aptos-labs/aptos-core/pull/15283 so we just need to test such some publish scenarios (with real db)

msmouse replied:
many tests fail with the panic.. I need to debug today to see how to enforce the assumption better. 

none the less I think we need to trigger some module republishing in a smoke test

igor-aptos replied:
@msmouse does it need to be a smoke test? we can have executor benchmark lib unit test with publishing with 3 lines of code

zekun000 replied:
it's hard to construct some complex scenario in smoke test because most of the time a block only has one txn (block produces too fast). it's easier to use executor + db alone to reproduce. I thought we already have module publish in smoke tests

igor-aptos replied:
```

    #[test]
    fn test_module_publish_txns() {
        AptosVM::set_num_shards_once(1);
        AptosVM::set_concurrency_level_once(4);
        AptosVM::set_processed_transactions_detailed_counters();
        test_generic_benchmark::<AptosVMBlockExecutor>(
            Some(TransactionTypeArg::PublishPackage),
            true,
        );
    }

this doesn't call on them afterwards though
```

msmouse replied:
we need to repeatedly publish the same module and the module must be being used so it resides in cache, is that simple to implement?

igor-aptos replied:
Should be, let me see

igor-aptos replied:
here is the republish and call mixed workload - #15292, and it fails.

you can use it to confirm/test this fix.

msmouse replied:
my paranoid panic pr landed, btw

gelash replied:
Shouldn't we just remove this assumption or deal at the storage layer wherever it causes problems? Not just for modules, but in general it would be great if MoveVM/BlockSTM could avoid unnecessary reads - of course we could ignore for conflicts and fake for storage, like here, but that's uglier and error prone from here?

georgemitenkov replied:
I think we should really drain readsets captured by execution then? We have all the info, whatever DB is doing with the reads, we already know one layer above.

msmouse replied:
yeah this is just for very short term mitigation. Going forward two things will happen:
1. the smt update will happen off the critical path, so we don't need to eager- read the proofs before the write set is formed. 
But 2. when we create the "hot tier" of state storage, we need to move state items between the hot tier and cold tier deterministically, that's where knowing precisely what gets read (for example, speculative reads that don't actually happen in the end need to be excluded) is REQUIRED. -- I'm gonna need to get the read set from the blockStm directly.

igor-aptos replied:
we have deterministic `ReadWriteSummary` for each transaction, produced by BlockSTM - which we use for determining when block gas limit has been hit.

We can probably provide that/variant of that to storage. For storage to use it - would it be required to add it to WriteSet, or?

msmouse replied:
I think making it part of the BlockEndInfo is proper.

igor-aptos replied:
but then BlockEndInfo would be read set for the whole block, which might be large? but that might be a good place for it.

msmouse replied:
yesh that's what I want. the trees like "bathcing effect"

gelash replied:
@msmouse that sounds great, one note is that we may have partial reads in the future, such as an existence check (can always be interpreted as a full read in lower layers, but in case it makes any difference)

gelash commented on aptos-move/block-executor/src/code_cache.rs:
```diff
@@ -136,44 +136,39 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> ModuleCache
+                    state
+                        .captured_reads
+                        .borrow_mut()
+                        .capture_global_cache_read(key.clone(), module.clone());
```
it would be a total overkill here but I wonder if we can do RAAI style little struct that captures things on drop, to make sure different paths of getting things all get recorded in captured reads (not just for modules). But I suppose we don't have such complex flows anywhere else.

georgemitenkov commented:
@gelash @zekun000 @igor-aptos Update:
- I moved read-before-write for modules from executor to VM. This way we can fail at a single txn, rather than returning a panic error (fallback/panic/block skip).
- Added @igor-aptos's test: panics without the fix, and logs invariant violations. With the fix - all good.

Let me know if changes to  read-before-write look good to you. Will merge on Sunday.
